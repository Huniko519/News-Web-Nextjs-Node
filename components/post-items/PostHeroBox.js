import React from 'react';
import PropTypes from 'prop-types';
import PostItemMedia from './PostItemMedia';
import PostItemTitle from './PostItemTitle';
import PostItemLabel from './PostItemLabel';
import Qoute from '../../utils/Qoute';

/**
 * Display a large image with the headline and exerpt
 * ╔══════════╤══════════════╗
 * ║ Headline │              ║
 * ║          │    Image     ║
 * ║ Excerpt  │              ║
 * ╚══════════╧══════════════╝
 * @param {post: {...postData}} props
 */
const PostHeroBox = (props) => {
  const { post, showAuthor, showQuote } = props;
  const isSponsored = post.type === 'post_sponsored';
  return (
    <div className="inews__post inews__post-herobox" key={post.id} data-post-id={post.id}>
      <div className={`inews__post-herobox__content ${(showAuthor && post.template === 'opinion') ? 'inews__post-herobox_space' : ''}`}>
        <div>
          {post.template === 'opinion' && showQuote && <Qoute size="medium" />}
          <PostItemTitle post={post} />
          <PostItemLabel post={post} />
          {isSponsored && (
            <span className="inews__post__sponsored-label">
              Promoted Content
            </span>
          )}
        </div>
        <div>
          {showAuthor && post.template === 'opinion' && (
            <div className="inews__post-herobox_author">
              <PostItemMedia
                className="inews__post-herobox_author__media"
                imageSize="small-16:9"
                width={230}
                height={130}
                layout="fixed"
                showAuthor
                placeholder="/static/images/placeholder/placeholder-84x84.png"
                post={post}
                linkToAuthor
              />
              <p className="inews__post__badge">
                <a href={post['co-authors'][0].link}>
                  {post['co-authors'][0].display_name}
                </a>
              </p>
            </div>
          )}
          <p
            className={`inews__post-herobox_description ${(post.template === 'opinion') ? 'hidden' : ''}`}
            dangerouslySetInnerHTML={{
              __html: post.excerpt.rendered,
            }}
          />
        </div>
      </div>
      <div>
        {/* div is needed for positioning context for video icon to prevent edge cases with grid */}
        <PostItemMedia
          className="inews__post-herobox__media inews__post-herobox__media--1-1"
          imageSize="medium-1:1"
          width={1}
          height={1}
          layout="responsive"
          placeholder="/static/images/placeholder/placeholder-84x84.png"
          post={post}
        />
        <PostItemMedia
          className="inews__post-herobox__media inews__post-herobox__media--16-9"
          imageSize="medium-16:9"
          width={16}
          height={9}
          layout="responsive"
          placeholder="/static/images/placeholder/placeholder-640x360.png"
          post={post}
        />
      </div>
    </div>
  );
};

/* istanbul ignore next */
PostHeroBox.propTypes = {
  post: PropTypes.objectOf(PropTypes.any).isRequired,
  showAuthor: PropTypes.bool,
  showQuote: PropTypes.bool,
};
PostHeroBox.defaultProps = {
  showAuthor: false,
  showQuote: false,
};

export default PostHeroBox;
