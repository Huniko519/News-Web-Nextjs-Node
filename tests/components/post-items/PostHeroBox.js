import React from 'react';
import PostHeroBox from '../../../components/post-items/PostHeroBox';

test('renders the snapshot', () => {
  const post = {
    id: 1,
    title: {
      rendered: 'title',
    },
    excerpt: {
      rendered: 'excerpt',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article',
  };
  const postHeroBox = shallow(<PostHeroBox post={post} />);
  expect(postHeroBox).toMatchSnapshot();
});

it('should output with the correct class on the root element', () => {
  const post = {
    id: 1,
    title: {
      rendered: 'title',
    },
    excerpt: {
      rendered: 'excerpt',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article',
  };
  const postHero = shallow(<PostHeroBox post={post} />);
  expect(postHero.hasClass('inews__post')).toBe(true);
  expect(postHero.hasClass('inews__post-herobox')).toBe(true);
});

it('should call the <PostItemMedia /> component with the correct props', () => {
  const post = {
    id: 1,
    title: {
      rendered: 'title',
    },
    excerpt: {
      rendered: 'excerpt',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article',
  };

  const postHero = shallow(<PostHeroBox post={post} />);
  const postItemMedia = postHero.find('PostItemMedia');
  expect(postItemMedia.length).toBe(2);
  expect(postItemMedia.at(0).prop('className')).toBe('inews__post-herobox__media inews__post-herobox__media--1-1');
  expect(postItemMedia.at(0).prop('imageSize')).toBe('medium-1:1');
  expect(postItemMedia.at(0).prop('width')).toBe(1);
  expect(postItemMedia.at(0).prop('height')).toBe(1);
  expect(postItemMedia.at(0).prop('layout')).toBe('responsive');
  expect(postItemMedia.at(0).prop('post')).toMatchObject(post);

  expect(postItemMedia.at(1).prop('className')).toBe('inews__post-herobox__media inews__post-herobox__media--16-9');
  expect(postItemMedia.at(1).prop('imageSize')).toBe('medium-16:9');
  expect(postItemMedia.at(1).prop('width')).toBe(16);
  expect(postItemMedia.at(1).prop('height')).toBe(9);
  expect(postItemMedia.at(1).prop('layout')).toBe('responsive');
  expect(postItemMedia.at(1).prop('post')).toMatchObject(post);
});

it('should call the <PostItemTitle /> component with the correct props', () => {
  const post = {
    id: 1,
    title: {
      rendered: 'title',
    },
    excerpt: {
      rendered: 'excerpt',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article',
  };

  const postHero = shallow(<PostHeroBox post={post} />);
  const postItemTitle = postHero.find('PostItemTitle');
  expect(postItemTitle.length).toBe(1);
  expect(postItemTitle.prop('post')).toMatchObject(post);
});

it('should output a Sponsored label if post.type === post_sponsored', () => {
  const post = {
    id: 1,
    type: 'post_sponsored',
    title: {
      rendered: 'title',
    },
    excerpt: {
      rendered: 'excerpt',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article',
  };
  const postHero = shallow(<PostHeroBox post={post} />);
  const label = postHero.find('span.inews__post__sponsored-label');
  expect(label.length).toBe(1);
  expect(label.text()).toBe('Promoted Content');
});

it('should output a author and author image if post.template === opinion and showAuthor prop is true', () => {
  const post = {
    'id': 1,
    'template': 'opinion',
    'co-authors': [
      {
        display_name: 'Andrew Fisher',
      },
    ],
    'title': {
      rendered: 'title',
    },
    'excerpt': {
      rendered: 'excerpt',
    },
    'breadcrumbs': [
      { name: 'News' },
    ],
    'link': 'https://inews-article.com/test-article',
  };
  const postHero = shallow(<PostHeroBox post={post} showAuthor />);
  const label = postHero.find('.inews__post-herobox_author .inews__post__badge a');
  expect(label.length).toBe(1);
  expect(label.text()).toBe('Andrew Fisher');

  const postItemMedia = postHero.find('.inews__post-herobox_author__media');
  expect(postItemMedia.length).toBe(1);
  expect(postItemMedia.prop('imageSize')).toBe('small-16:9');
  expect(postItemMedia.prop('width')).toBe(230);
  expect(postItemMedia.prop('height')).toBe(130);
  expect(postItemMedia.prop('layout')).toBe('fixed');
  expect(postItemMedia.prop('placeholder')).toBe('/static/images/placeholder/placeholder-84x84.png');
  expect(postItemMedia.prop('post')).toMatchObject(post);
  expect(postItemMedia.prop('showAuthor')).toBe(true);
});

it('should not output a author and author image if post.template !== opinion or showAuthor prop is false', () => {
  const post = {
    id: 1,
    title: {
      rendered: 'title',
    },
    excerpt: {
      rendered: 'excerpt',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article',
  };
  const postHero = shallow(<PostHeroBox post={post} />);
  const label = postHero.find('.inews__post-herobox_author .inews__post__badge a');
  expect(label.length).toBe(0);
  const postItemMedia = postHero.find('.inews__post-herobox_author__media');
  expect(postItemMedia.length).toBe(0);
});
