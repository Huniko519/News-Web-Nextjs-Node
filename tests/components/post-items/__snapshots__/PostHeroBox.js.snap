// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the snapshot 1`] = `
<div
  className="inews__post inews__post-herobox"
  data-post-id={1}
  key="1"
>
  <div
    className="inews__post-herobox__content "
  >
    <div>
      <PostItemTitle
        post={
          Object {
            "breadcrumbs": Array [
              Object {
                "name": "News",
              },
            ],
            "excerpt": Object {
              "rendered": "excerpt",
            },
            "id": 1,
            "link": "https://inews-article.com/test-article",
            "title": Object {
              "rendered": "title",
            },
          }
        }
      />
      <PostItemLabel
        post={
          Object {
            "breadcrumbs": Array [
              Object {
                "name": "News",
              },
            ],
            "excerpt": Object {
              "rendered": "excerpt",
            },
            "id": 1,
            "link": "https://inews-article.com/test-article",
            "title": Object {
              "rendered": "title",
            },
          }
        }
      />
    </div>
    <div>
      <p
        className="inews__post-herobox_description "
        dangerouslySetInnerHTML={
          Object {
            "__html": "excerpt",
          }
        }
      />
    </div>
  </div>
  <div>
    <PostItemMedia
      className="inews__post-herobox__media inews__post-herobox__media--1-1"
      height={1}
      imageSize="medium-1:1"
      layout="responsive"
      linkToAuthor={false}
      placeholder="/static/images/placeholder/placeholder-84x84.png"
      post={
        Object {
          "breadcrumbs": Array [
            Object {
              "name": "News",
            },
          ],
          "excerpt": Object {
            "rendered": "excerpt",
          },
          "id": 1,
          "link": "https://inews-article.com/test-article",
          "title": Object {
            "rendered": "title",
          },
        }
      }
      postItemType=""
      showAuthor={false}
      width={1}
    />
    <PostItemMedia
      className="inews__post-herobox__media inews__post-herobox__media--16-9"
      height={9}
      imageSize="medium-16:9"
      layout="responsive"
      linkToAuthor={false}
      placeholder="/static/images/placeholder/placeholder-640x360.png"
      post={
        Object {
          "breadcrumbs": Array [
            Object {
              "name": "News",
            },
          ],
          "excerpt": Object {
            "rendered": "excerpt",
          },
          "id": 1,
          "link": "https://inews-article.com/test-article",
          "title": Object {
            "rendered": "title",
          },
        }
      }
      postItemType=""
      showAuthor={false}
      width={16}
    />
  </div>
</div>
`;
